asulibraries/islandora-repo
# Tsv merge

This feature branch includes 
1. a script to merge the a legacy repo CSV export* with a TSV that was exported via MARC ETD and converted into this proprietary format -- applying logic on which fields to overwrite from the MARC TSV file.
2. a set of migration scripts for ingesting the resultant file from the merge script
3. adjusted the parsing of migration script to always use delimiter " || " and to also import the Note type -- applied this also to the migration scripts for c269 since it is used as a copy-template in making new migration scripts.**

---
 ( * )- the c7_merged_v2 file had one item (Item ID = 43976) where the description field value contained a double-quote character. We manually edited (changed {G\""ortler} to {G'ortler}) that row before merging and migration the result. 
 ( ** )- the adjustment to the migration script requires that the default separation character for the legacy repo CSV export be changed to use the " || " to separate all multi-items.


The merge script also has a helper unit `web/modules/custom/asu_migrate/handle_id_map.php` which provides the handle mapping between weird handle values that appeared in the TSV export which needed to correlate to real "Item ID" values.

### To test:
The migration scripts must be imported:
```
web/modules/custom/asu_migrate/config/install/migrate_plus.migration.c7_node.yml 
web/modules/custom/asu_migrate/config/install/migrate_plus.migration.c7_media.yml 
web/modules/custom/asu_migrate/config/install/migrate_plus.migration.c7_file.yml 
```
The files needed to merge are NOT a part of the github codebase, so c7_merged_v2.csv, att_file_7_v2.csv, and marc_etd_20210715_rev.tsv should be either mapped or copied into the working folder. This should be able to process the entire 7000+ TSV items and the entire CSV file in one pass (but this used to run out of memory in earlier development versions, so the parameters for offset (-o) and "how many rows to process" (-n)

#### The path to the TSV file, the CSV file, and where to save the result are required to run. To see the merge scripts options and help:
```
cd /var/www/html/drupal/web/modules/custom/asu_migrate
php merge_repo_and_tsv.php --help
```

#### to run without xdebug:
From the terminal, run the merge_repo_and_tsv.php script:
```
cd /var/www/html/drupal/web/modules/custom/asu_migrate
php merge_repo_and_tsv.php -tsv=marc_etd_20210715_rev.tsv -csv=c7_merged_v2.csv -out=c7_merged_v3.csv -n=20000 -o=0
```

#### to run with xdebug:
```
cd /var/www/html/drupal/web/modules/custom/asu_migrate
php -dxdebug.mode=debug -dxdebug.client_host=10.0.2.2 -dxdebug.client_port=9003 -dxdebug.idekey="YOUR_IDE_NAME_HERE" -dxdebug.start_with_request=yes ./merge_repo_and_tsv.php -tsv=marc_etd_20210715_rev.tsv -csv=c7_merged_v2.csv -out=c7_merged_v3.csv -n=20000 -o=0
```

After this, the resultant "c7_merged_v3.csv" file is used by the c7_node migration script.

Please let me know if there are any questions since this feature branch is a serious one-off.
